### PR TITLE
Adding additional Init Param to skip certain endpoints with the filter

### PR DESCRIPTION
**Pull request by BHoTasktop**
implemented the filter in the same way as keycloak's filter on pattern matching.

### CODE REVIEW

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 1]
```diff
@@ -38,15 +39,29 @@
 	private static final String HEADER_X_FORWARDED_HOST = "X-Forwarded-Host";
 	private static final String HEADER_ORIGIN = "Origin";
 	private static final String HEADER_REFERER = "Referer";
+	private static final String SKIP_PATTERN_PARAM = "cors.config.skipPattern";
+	Pattern skipPattern;
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: skipPattern should be private

**Comment by BHoTasktop** [comment]
Reopen PR to trigger Travis-ci

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 2]
```diff
@@ -38,15 +39,29 @@
 	private static final String HEADER_X_FORWARDED_HOST = "X-Forwarded-Host";
 	private static final String HEADER_ORIGIN = "Origin";
 	private static final String HEADER_REFERER = "Referer";
+	private static final String SKIP_PATTERN_PARAM = "cors.config.skipPattern";
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: Consider more intuitive naming, e.g.

 path-exclusion-pattern

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 3]
```diff
@@ -38,15 +39,29 @@
...
+	private Pattern skipPattern;
 
 	@Override
 	public void init(FilterConfig filterConfig) throws ServletException {
-		// nothing to do
+		String skipPatternDefinition = filterConfig.getInitParameter(SKIP_PATTERN_PARAM);
+
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: no need for blank lines here

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 4]
```diff
@@ -38,15 +39,29 @@
 	private static final String HEADER_X_FORWARDED_HOST = "X-Forwarded-Host";
 	private static final String HEADER_ORIGIN = "Origin";
 	private static final String HEADER_REFERER = "Referer";
+	private static final String SKIP_PATTERN_PARAM = "cors.config.skipPattern";
+	private Pattern skipPattern;
 
 	@Override
 	public void init(FilterConfig filterConfig) throws ServletException {
-		// nothing to do
+		String skipPatternDefinition = filterConfig.getInitParameter(SKIP_PATTERN_PARAM);
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: consider renaming to pathExclusionPattern for the variable and the field

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 5]
```diff
@@ -38,15 +39,29 @@
...
+		}
+
 	}
 
 	@Override
 	public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
 			throws IOException, ServletException {
+
+		if (shouldSkip((HttpServletRequest)request)) {
+			chain.doFilter(request, response);
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: there is some redundancy here, consider rewriting to something like:

 if (!requestPathIsExcluded(request)) {
    ... check request headers
 }
 chain.doFilter(...)

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 6]
```diff
@@ -56,6 +71,15 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
 		chain.doFilter(request, response);
 	}
 
+	private boolean shouldSkip(HttpServletRequest request) {
+		if (skipPattern == null) {
+			return false;
+		}
+
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: remove blank line

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 7]
```diff
@@ -56,6 +71,15 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
...
+		String requestPath = request.getRequestURI().substring(request.getContextPath().length());
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: the path here is not decoded, i.e. could take many forms, for example:

* /some+path
* /some%20path

consider using the decoded path here instead

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java [hunk 8]
```diff
@@ -38,15 +39,29 @@
...
+	private static final String SKIP_PATTERN_PARAM = "cors.config.skipPattern";
+	private Pattern skipPattern;
 
 	@Override
 	public void init(FilterConfig filterConfig) throws ServletException {
-		// nothing to do
+		String skipPatternDefinition = filterConfig.getInitParameter(SKIP_PATTERN_PARAM);
+
+		if (skipPatternDefinition != null) {
+			skipPattern = Pattern.compile(skipPatternDefinition, Pattern.DOTALL);
```
Comment (greensopinion) [inline src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: Regular expressions are super flexible, but also hard to get right.  I wonder if we should consider instead using path prefixes here instead of regular expressions, for example:

* /a/path/prefix -> would map to any path starting with /a/path/prefix, such as /a/path/prefix/to-any-file

Also we should consider allowing the init param to contain multiple paths.  Have you considered a comma-delimited list of values?  The Spring framework uses commas, semicolons and whitespace (any are allowed).

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java [hunk 9]
```diff
@@ -165,6 +170,16 @@ public void doFilterRejectsRequestWithRefererWithoutHost() throws IOException, S
 		expectForbidden();
 	}
 
+	@Test
+	public void doFilterAcceptsRequestWithEmptyRefererWhenPathIsSkipped() throws IOException, ServletException {
+		doReturn("^.*").when(config).getInitParameter(any());
```
Comment (greensopinion) [inline src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: any() is not specific enough, instead provide the parameter name here

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java [hunk 10]
```diff
@@ -165,6 +170,16 @@ public void doFilterRejectsRequestWithRefererWithoutHost() throws IOException, S
 		expectForbidden();
 	}
 
+	@Test
+	public void doFilterAcceptsRequestWithEmptyRefererWhenPathIsSkipped() throws IOException, ServletException {
```
Comment (greensopinion) [inline src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: instead of "skipped" consider consistently using "excluded"

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java [hunk 11]
```diff
@@ -165,6 +170,16 @@ public void doFilterRejectsRequestWithRefererWithoutHost() throws IOException, S
...
+		doReturn("^.*").when(config).getInitParameter(any());
+		filter.init(config);
+		mockHeader(HTTP_HEADER_REFERER, "http://a-host/some/path");
+		filter.doFilter(request, response, chain);
+		verify(chain).doFilter(request, response);
+		verifyNoMoreInteractions(chain);
+	}
+
```
Comment (greensopinion) [inline src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: should have more tests here to cover edge cases, e.g.

* init parameter is provided but empty
* init parameter is provided but invalid
* init parameter is provided and paths are excluded, but request path is not excluded and headers are checked
* init parameter is provided and paths are excluded, but request path is excluded and headers are not checked
* with a context path, without a context path
* with path having percent-encoded characters
* ...

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 2 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: done

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 3 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: done

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 4 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: done

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 5 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: done

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 6 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: done

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 7 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: will implement.

File: src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java (hunk 8 above)
Comment (BHoTasktop) [reply src/main/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilter.java]: There were two reasons behind the Regex pattern.
1. the endpoint we are allowing through takes forms like this. /auth/realms/{Tasktop}/broker/{saml}/endpoint where the two path levels are configured through keycloak. 
2. I have been told that keycloak doesn't handle cores very well, ergo we are filtering cors requests at the tomcat level. Having just a prefix (ie. /auth/realms/Tasktop/broker/) may open up unwanted endpoints in this case.

With the 2 constrains I think allowing multiple path would solve these issue. Will implement.

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java (hunk 9 above)
Comment (BHoTasktop) [reply src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: done

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java (hunk 10 above)
Comment (BHoTasktop) [reply src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: done

File: src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java (hunk 11 above)
Comment (BHoTasktop) [reply src/test/java/com/tasktop/servlet/cors/CorsHeaderScrutinyServletFilterTest.java]: understood, will add more test cases.
